pkg/builder/openapi.go

```diff
@@ -45,7 +45,46 @@ type openAPI struct {
 
 // BuildOpenAPISpec builds OpenAPI spec given a list of webservices (containing routes) and common.Config to customize it.
 func BuildOpenAPISpec(webServices []*restful.WebService, config *common.Config) (*spec.Swagger, error) {
-	o := openAPI{
+	o := newOpenAPI(config)
+
+	err := o.init()
```

Comment:
Nit: It's a private struct so it doesn't matter that much, but you've created an implicit dependency in the order in which these functions need to be called. Can you call `init` in `newOpenAPI`, since you already have a sort of constructor? 

`finishSpec` is also suffering from these syndroms, you need to call it before you can get the `o.swagger`, if you forget to call it then, who knows what will happen. Can we "finishSpec" before (like in init) or do we need all the data before? Then you would have:

```
o, err := newOpenAPIConfig(config)
// handle err
o.addPath(webservices)
return o.Swagger() // But this function should be const.
```

Reply:
Can definitely call init in newOpenAPI, but we do need the spec to be built before calling finishSpec (it contains things like a postProcessSpec function, which is defined by the openapiconfig)

Reply:
Would a function like this be any better?
```
func createSpec(config common.OpenAPIConfig, buildSwaggerFunc func(o *openAPI)) *spec.Swagger {
o := openAPI{
...
}
o.init()
buildSwaggerFunc(o)
o.finish()
return o.swagger
}
```
(plus some error handling)

Reply:
~~Meh, I don't know if you have many options here. It looks like the `openAPI` struct is having too many side-effects already. Let's see how we can make things better without changing everything ...~~

~~I'm tempted to think that these methods in general should be free-functions instead (because they don't really maintain a cohesive state for the struct anyway).~~

Yeah, maybe you should leave it as it is. :-)

Reply:
Okay, and I added a comment documenting the requirement to call finish before returning the spec.